### PR TITLE
semantics metrics assertion, part 1: auxiliary GomegaMatcher-s

### PR DESCRIPTION
**Pull request by Ray-Eldath**
(feat) GomegaMatcher for Array & Slice: AllPassed and AnyFailed, tested & documented, in package matcher. **Go for `container_test.go` to see how to use these two convienent matcher when working with array or slice.**
(feat) GomegaMatcher for metrics checking with PromQL, documented, in package matcher.
(feat) GomegaMatcher for panics with message printed (and an omissible parameter "name"), documented, in package matcher.
(feat) ConsistentlyWithMetrics for another set of defaults, tested, in package assertion.
(refactor) refactor ParameterizedTestCase with Golang builtin parsing method.
(refactor) refactor demonstrating tikv_test.go with newly added PromQL matcher.

### CODE REVIEW

File: pkg/ginkgo_helper/metrics.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	"github.com/onsi/gomega"
+	"github.com/onsi/gomega/format"
+	"github.com/pingcap/test-infra/sdk/resource"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+)
+
+func BuildPrometheusAPI(tc resource.TiDBCluster) (v1.API, error) {
+	url, err := tc.ServiceURL(resource.Prometheus)
```
Comment (cosven) [inline pkg/ginkgo_helper/metrics.go]: I think we should not import infra sdk in other pacakges.

File: pkg/ginkgo_helper/metrics.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	} else {
+		return format.Message(matcher.actual, fmt.Sprintf("to match the expected result of PromQL [%s] %s", matcher.description, matcher.promQL), "<empty Scalar or Vector>")
+	}
+}
+
+func init() {
+	gomega.SetDefaultConsistentlyDuration(3 * time.Hour)
+	gomega.SetDefaultConsistentlyPollingInterval(1 * time.Minute)
+	gomega.SetDefaultEventuallyTimeout(3 * time.Hour)
+	gomega.SetDefaultEventuallyPollingInterval(1 * time.Minute)
```
Comment (cosven) [inline pkg/ginkgo_helper/metrics.go]: Why change default values here?

File: testcase/oom/tikv_test.go [hunk 3]
```diff
@@ -11,12 +11,12 @@ import (
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 
-	qe_metrics "github.com/PingCAP-QE/metrics-checker/pkg/metrics"
 	_ "github.com/go-sql-driver/mysql"
+	infraresource "github.com/pingcap/test-infra/sdk/resource"
```
Comment (cosven) [inline testcase/oom/tikv_test.go]: IMO, `infra_resource` has better readability 🤔.

File: pkg/ginkgo_helper/metrics.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+func PromQL(description string, query string, api *v1.API) PromQLInstance {
+	return PromQLInstance{
+		description: description,
+		query:       query,
+		api:         api,
+	}
+}
+
+type promQLEvaluatedMatcher struct {
```
Comment (cosven) [inline pkg/ginkgo_helper/metrics.go]: Can we seperate the gomega logic from promQL check logic?

File: testcase/oom/tikv_test.go (hunk 3 above)
Comment (Ray-Eldath) [reply testcase/oom/tikv_test.go]: but will makes GoLand unhappy 😂😂

File: pkg/ginkgo_helper/metrics.go (hunk 1 above)
Comment (Ray-Eldath) [reply pkg/ginkgo_helper/metrics.go]: We could let user pass in the result of `ServiceURL` directly.

File: pkg/ginkgo_helper/metrics.go (hunk 2 above)
Comment (Ray-Eldath) [reply pkg/ginkgo_helper/metrics.go]: default value is *"`Eventually` will poll every 10 milliseconds for up to 1 second and `Consistently` will monitor every 10 milliseconds for up to 100 milliseconds"* which is undoubtedly too frequent for a metrics checking, and the timeout is too short as we need to moniter the status of a cluster in a relatively long time as well.

File: pkg/ginkgo_helper/metrics.go (hunk 4 above)
Comment (Ray-Eldath) [reply pkg/ginkgo_helper/metrics.go]: the main PromQL logic is only a small function, roughly 15 lines of code (`pkg/ginkgo_helper/metrics.go:74`). I think keep the function inplace could make it more compact and more readable.

File: pkg/ginkgo_helper/metrics.go (hunk 2 above)
Comment (cosven) [reply pkg/ginkgo_helper/metrics.go]: Can we only set default value for metrics-checking? It seems that this PR changes the  global default value.

File: testcase/oom/tikv_test.go (hunk 3 above)
Comment (cosven) [reply testcase/oom/tikv_test.go]: I see😂. I checked the blog [package-names](https://blog.golang.org/package-names).

Maybe we can use name `resource` directly?

File: pkg/ginkgo_helper/metrics.go (hunk 4 above)
Comment (cosven) [reply pkg/ginkgo_helper/metrics.go]: LGTM.

File: pkg/ginkgo_helper/collection.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+type anyShouldNotMatcher struct {
+	assertion types.GomegaMatcher
+	evaluated bool
+
+	actual           interface{}
+	successElements  []reflect.Value
+	successPositions []int
+}
+
+func AnyShouldFailed(assertion types.GomegaMatcher) *anyShouldNotMatcher {
```
Comment (cosven) [inline pkg/ginkgo_helper/collection.go]: What does `AnyShouldFailed` mean?

File: pkg/ginkgo_helper/metrics.go (hunk 1 above)
Comment (cosven) [reply pkg/ginkgo_helper/metrics.go]: Yeah!

File: pkg/ginkgo_helper/collection.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	"reflect"
+
+	"github.com/onsi/gomega/format"
+	"github.com/onsi/gomega/types"
+	"github.com/pingcap/errors"
+)
+
+type allShouldMatcher struct {
+	assertion types.GomegaMatcher
+	evaluated bool
```
Comment (cosven) [inline pkg/ginkgo_helper/collection.go]: It seems  the attribute `evaluated` is not used.

File: pkg/ginkgo_helper/metrics.go (hunk 2 above)
Comment (Ray-Eldath) [reply pkg/ginkgo_helper/metrics.go]: seems not supportted by gomega. Or we could make our own version of `Consistently` (maybe `ConsistentlyOnMetrics`...?) that works as a pure delegate of the original `Consistently` with longer interval & duration passed in.

**Comment by Ray-Eldath** [comment]
@cosven problems addressed, PTAL.

**Comment by Ray-Eldath** [comment]
@cosven now its completely dual, ptal. hope this is the last shot, then we can work on `endless`. 😉

File: pkg/ginkgo_helper/metrics.go (hunk 2 above)
Comment (cosven) [reply pkg/ginkgo_helper/metrics.go]: Maybe we can define some varaibles and case can use them in metrics matcher?

File: pkg/ginkgo_helper/metrics.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	"time"
+
+	"github.com/onsi/gomega"
+	"github.com/onsi/gomega/format"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+)
+
+func BuildPrometheusAPI(prometheusURL url.URL) (v1.API, error) {
```
Comment (cosven) [inline pkg/ginkgo_helper/metrics.go]: How about change the name to `NewPrometheusClient`?

File: pkg/ginkgo_helper/panic.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+func (p *panicWithOutput) FailureMessage(actual interface{}) (message string) {
+	return p.delegate.FailureMessage(actual)
+}
+
+func (p *panicWithOutput) NegatedFailureMessage(actual interface{}) (message string) {
+	return p.delegate.NegatedFailureMessage(actual)
+}
+
+func PanicWithOutput(name ...string) types.GomegaMatcher {
```
Comment (cosven) [inline pkg/ginkgo_helper/panic.go]: We should add some docs for these dsl lator.

**Comment by Ray-Eldath** [comment]
@cosven all done. :wink:

**Review by cosven — APPROVED** [review]
LGTM

**Pull request closed without merging**